external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/KafkaSpoutCommitTest.java

```diff
@@ -0,0 +1,145 @@
+        try (SimulatedTime simulatedTime = new SimulatedTime()) {
+            setupSpout(Collections.singleton(partition));
+            Map<TopicPartition, List<ConsumerRecord<String, String>>> records = new HashMap<>();
+            List<ConsumerRecord<String, String>> recordsForPartition = new ArrayList<>();
+            // Offsets emitted are 0,1,2,3,4,...,8,9
+            for (int i = 0; i < 5; i++) {
+                recordsForPartition.add(new ConsumerRecord(partition.topic(), partition.partition(), i, "key", "value"));
+            }
+            for (int i = 8; i < 10; i++) {
+                recordsForPartition.add(new ConsumerRecord(partition.topic(), partition.partition(), i, "key", "value"));
+            }
+            records.put(partition, recordsForPartition);
+
+            when(consumerMock.poll(anyLong()))
+                    .thenReturn(new ConsumerRecords(records));
+
+            for (int i = 0; i < recordsForPartition.size(); i++) {
+                spout.nextTuple();
+            }
+
+            ArgumentCaptor<KafkaSpoutMessageId> messageIds = ArgumentCaptor.forClass(KafkaSpoutMessageId.class);
+            verify(collectorMock, times(recordsForPartition.size())).emit(anyObject(), anyObject(), messageIds.capture());
+
+            for (KafkaSpoutMessageId messageId : messageIds.getAllValues()) {
+                spout.ack(messageId);
+            }
+
+            // all tuples emitted are pending commit
+            assertEquals(recordsForPartition.size(), spout.getNumUncommittedOffsets());
+
+            // Advance time and then trigger first call to kafka consumer commit; the commit will progress till offset 4
+            Time.advanceTime(KafkaSpout.TIMER_DELAY_MS + offsetCommitPeriodMs);
+            when(consumerMock.poll(anyLong()))
+                    .thenReturn(new ConsumerRecords(Collections.EMPTY_MAP));
+            spout.nextTuple();
+
+            ArgumentCaptor<Map> committedOffsets=ArgumentCaptor.forClass(Map.class);
+            InOrder inOrder = inOrder(consumerMock);
+            inOrder.verify(consumerMock).commitSync(committedOffsets.capture());
+            inOrder.verify(consumerMock).poll(anyLong());
+
+            //verify that Offset 4 was last committed offset and 5 offsets were committed
+            //the offset void should be bridged in the next commit
+            assertTrue(committedOffsets.getValue().containsKey(partition));
+            assertEquals(((OffsetAndMetadata) (committedOffsets.getValue().get(partition))).offset(), 4);
+            assertEquals(2, spout.getNumUncommittedOffsets());
+
+            //Trigger second kafka consumer commit
+            reset(consumerMock);
+            when(consumerMock.poll(anyLong()))
+                    .thenReturn(new ConsumerRecords(Collections.EMPTY_MAP));
+            Time.advanceTime(KafkaSpout.TIMER_DELAY_MS + offsetCommitPeriodMs);
+            spout.nextTuple();
+
+            committedOffsets=ArgumentCaptor.forClass(Map.class);
```

Comment:
Nit: You can use the Mock annotation to put generic parameters on this. See https://static.javadoc.io/org.mockito/mockito-core/2.8.9/org/mockito/Mockito.html#mock_annotation